semver.md

```diff
@@ -76,8 +76,8 @@ is incremented after this release is dependent on this public API and how it
 changes.
 
 1. Patch version Z (x.y.Z | x > 0) MUST be incremented if only backwards
-compatible bug fixes are introduced. A bug fix is defined as an internal
-change that fixes incorrect behavior.
+compatible changes are made. Such changes include but are not limited to:
+internal changes that fix incorrect behavior, or refactorings.
```

Comment:
By definition, a refactoring doesn't change the behavior at all. It probably doesn't make sense to release a new public version of a package that _only_ includes refactorings. If you did, it should be the same version since the behavior didn't change.

However, many folks conflate "refactoring" with "bug fixing while cleaning the code" in which case "bug fix" covers the scenario.

Reply:
I do not believe it is possible to refactor code without having an impact on performance or size of the package, either of which carry some risk.  What about security critical code that uses your interface implementation?  Would you want to lie to them?  Is it even legal under SemVer to release different packages with the same name, decorations (arch/debug/codecoverage, etc) and version number on them?  I would hope not.

Reply:
One good reason to release a refactored version is to test your assumptions that it is indeed a plug-in replacement for the earlier version and allow some back-time in the consumer community prior to proceeding with new feature development.

Reply:
>3. Once a versioned package has been released, the contents of that version MUST NOT be modified. Any modifications MUST be released as a new version.

Reply:
> By definition, a refactoring doesn't change the behavior at all.

@Haacked I beg to differ, improved performance is a change in behavior. Yet not a "functional" change. A new release based on a refactoring that only improves performance is perfectly valid.

> in which case "bug fix" covers the scenario.

Personally I don't like the terminology of 'bugfix' since it ties the entire spec to software development. Alternative terminology that is more open is preferred imo.

Reply:
> It probably doesn't make sense to release a new public version of a package that only includes refactorings. If you did, it should be the same version since the behavior didn't change.

@Haacked the purpose of versioning goes beyond just functional changes. From a DevOps perspective I would want each version out there to be separately versioned. A refactoring (though it should not) might  introduce a new bug. And when someone hits that bug I'd want to know in which version of the code that happened.

One might even consider that a bad instruction in the `readme` for getting-started is a reason for a new patch version with the fixed instruction. You probably wouldn't do that, and me neither probably, but it is not up to the SemVer spec to be opinionated on that.

SemVer should not be opinionated, and it is up to the author to specify the exact scope of the version he or she stuck on a release, and there just is no one-size-fits-all for this.

Reply:
That last remark on "scope" should probably become a new FAQ entry.

Reply:
The definition of "refactoring" I use is the one Martin Fowler notes in the title of his book, [Refactoring: Improving the Design of Existing Code](https://amzn.to/2Hf4Mow). In [his bliki](https://martinfowler.com/bliki/DefinitionOfRefactoring.html), he notes,

> Refactoring (noun): a change made to the internal structure of software to make it easier to understand and cheaper to modify without changing its observable behavior.

Thus any change in behavior including performance improvements would not be a refactoring.

> either of which carry some risk

This resonates with me. It doesn't make sense to me that you'd ship a release with just refactorings (as defined by Fowler) because there shouldn't be any outwardly visible change. But my preference doesn't stop people from doing that. And if they do, there's always the risk they've incorrectly refactored the code and introduced a change.

Semver is about communicating risk. So it make sense to me to release a patch update in this case.

However, still not convinced by the specific proposed language. As noted elsewhere, you can add a new feature and it would still be a backward compatible change.

I think the spirit of the patch is that there's no outward functional change other than behavior is corrected (bug fix) or performance is improved. Are there other backward compatible changes that don't fall in that category?

Reply:
How about this?

MAJOR version when you make incompatible API changes,
MINOR version when you add to the API in a backwards-compatible manner, and
PATCH version when you make any other backwards compatible changes.

Reply:
And:

6. Patch version Z (x.y.Z | x > 0) MUST be incremented if only backwards compatible changes are introduced. This includes internal changes that fix incorrect behavior (bug fixes), but does not include any changes to the public API.

Reply:
@Haacked, Fowler is wrong in the strictest sense. It is not possible to modify source code, or the compiler for that matter, without making "observable changes". He's correct in the sense that the intended behavior, it's functionality, is not changed by refactoring. We can argue over whether a few nano-seconds here or there really constitutes a perf gain or not, but I have worked on systems where such differences are flagged by automation, trigger broader longer running test runs and must be signed-off on.  For some security conscious consumers, even a change in the content hash of the bits is reason for alarm.

Reply:
@jwdonahue 

> It is not possible to modify source code, or the compiler for that matter, without making "observable changes".

I don't think Fowler is wrong (if even possible at all, since definitions are almost always opinionated). How about replacing an `if-then` sequence by a `select`? "_to make it easier to understand and cheaper to modify_".

Reply:
@jwdonahue from a previous comment: if in a repo the docs are updated because of some faulty instruction in those docs. Some would consider it a bugfix, some wouldn't. It's up to the package owner to define whether the docs are in scope of the SemVer version number applied to the package. Or maybe he could choose to have 2 version numbers, 1 for the code, 1 for the docs.

Another example (to stay away from zebra's and books 😄): a components has a REST interface, but it also stores some data in a DB. Some users tend to directly access the DB besides the REST interface. Now if a SemVer version is applied, is that on the REST interface, or does that include the DB schema?

It's always up to the author of a package/component to define what is in scope and what not.